docs/root/configuration/http/http_conn_man/vhds.rst

```diff
@@ -9,6 +9,100 @@ will call to dynamically fetch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Virtual hosts in VHDS are identified by a combination of the name of the route
+configuration to which the virtual host belong as well as the host HTTP "host"
+header (authority for HTTP2) entry. Resources should be named as follows:
+
+<route configuration name>/<host entry>
+
+Note that matching should be done from right to left since a host entry cannot
+contain slashes while a route configuration name can. 
+
+Joining/Reconnecting to a management server
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When an Envoy instance forms a delta xDS connection to the xDS management
+server for the first time, it will receive a base configuration filtered down
+to a subset of routes that are likely useful to the Envoy instance. For
+example, these might be filtered down by routes that exist in the same
+application. The xDS management server will send this same collection of
+resources on reconnect.
+
+.. figure:: diagrams/delta_rds_connection.svg
+   :alt: Delta RDS connection
+
+Subscribing to Resources
+^^^^^^^^^^^^^^^^^^^^^^^^
+VHDS allows resources to be :ref:`subscribed <xds_protocol_delta_subscribe>` to
+using a :ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with the 
+:ref:`type_url <envoy_api_field_DeltaDiscoveryRequest.type_url>` set to 
+`type.googleapis.com/envoy.api.v2.route.VirtualHost` 
+and :ref:`resource_names_subscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_subscribe>` 
+set to a list of virtual host resource names for which it would like
+configuration. 
+
+If a route for the contents of a host/authority header cannot be resolved,
+the active stream is paused while a :ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>` is sent. 
+When a :ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>` is
+recieved where the host/authority header values exactly matches the 
+:ref:`aliases <envoy_api_field_resource.aliases>`,
+the route configuration is updated, the stream is resumed and processing of the
+filter chain continues.
+
+Each of the virtual hosts contained in the
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+resources field will be added to the route configuration maintained by Envoy.
+If Envoy's route configuration already contains a given virtual host, it will
+be overwritten but only if the updated virtual host is different from its
+current state. The management server needs to keep track of virtual hosts
+known to Envoy.
+
+If a virtual host is requested for which the management server does not know
+about, then the management server should respond with a 
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>` in which
+the :ref:`resources <envoy_api_field_DeltaDiscoveryResponse.resources>` entry
+for that virtual host has the :ref:`name <envoy_api_field_resource.name>` and
+:ref:`aliases <envoy_api_field_resource.aliases>` set to the requested host
```

Comment:
In this case, it seems like it should set `name` but would not need to set `aliases`.

Reply:
If the virtual host cannot be resolved, you know an alias that that virtual host could go by, but you don't the name of that virtual host since there is no virtual host.

Reply:
I think there should be no difference in content or semantics between the `name` and `aliases` fields -- if we didn't care about backward-compatibility, we could just change `name` to be a repeated field instead of adding the `aliases` field.  But even though they're split, the set of all values in both fields should be treated as a single list: all names in the list should be unique for the resource type, and a query for any one of the names should return the same object.

In this case, there is exactly one name that the client will be looking for.  So what's the value of putting that in the `aliases` field instead of in the `name` field?  And if we do put it in the `aliases` field, what will we put into the `name` field instead?

Reply:
Every virtual host has a definitive name that populates the name field and aliases are other names that that resource can go by. I'm not sure that we want to mix the meanings of the names and aliases. 

We also care about backwards compatibility of xDS, so I'm not sure we can change the API to combine names and aliases. 

A resource can have any name that the management server wants to assign it and Envoy doesn't care. However, it does care about the contents of the alias field. In the case of a VHDS resource not being found, we know that the requested resource can go by a name matching <routeConfigName>/<host>, but we don't know what name the management server assigned that resource since that resource could not be found.

Reply:
> Every virtual host has a definitive name that populates the name field and aliases are other names that that resource can go by. I'm not sure that we want to mix the meanings of the names and aliases.

Why would the meanings be different?  Are there any use-cases in which the client would care about the difference between them?

> We also care about backwards compatibility of xDS, so I'm not sure we can change the API to combine names and aliases.

This PR is exposing the `aliases` fields in the `Resource` message, which was not previously exposed, so we can define its semantics to be anything we want.  Why would this break backward compatibility any more than introducing the `aliases` field in the first place?

> A resource can have any name that the management server wants to assign it and Envoy doesn't care. However, it does care about the contents of the alias field. In the case of a VHDS resource not being found, we know that the requested resource can go by a name matching /, but we don't know what name the management server assigned that resource since that resource could not be found.

Envoy may not care about the name, but it's not true that no client cares.  The gRPC xDS client does in fact care about resource names; for example, in LDS and CDS requests, gRPC clients make non-wildcard requests, but for backward compatibility, it cannot assume that the server will honor the `resource_names` field in the request, so it looks for the specific resource name(s) that it asked for.

I think the right way to think about this case is that, because there is no match for the resource name that the client requested, the server is going to return the name that the client requested with no attached resource.  This tells the client that the requested resource does not exist.

Reply:
> Envoy may not care about the name, but it's not true that no client cares.

I am fairly certain that Envoy makes no assumptions about the format or the contents of resource names in xDS v2 api. In fact, keeping resource names opaque was one of the design considerations and is what led to introduction of aliases.

> I think the right way to think about this case is that, because there is no match for the resource name that the client requested, the server is going to return the name that the client requested with no attached resource. This tells the client that the requested resource does not exist.

This is close to how on-demand request misses are handled right now: the aliases field is used instead of the resource name field, however,

Reply:
> I am fairly certain that Envoy makes no assumptions about the format or the contents of resource names in xDS v2 api. In fact, keeping resource names opaque was one of the design considerations and is what led to introduction of aliases.

I think the change that I am proposing here breaks any of the existing assumptions.

The incremental xDS protocol has already established the fact that resources can have more than one name (all but one are called aliases instead of names, but they all serve the same function), and because ["[b]oth the names and aliases of resources should be checked in order to determine whether the entity in question has been subscribed to"](https://github.com/markdroth/envoy/blob/44c67d11c51a58797d05b77892c806281c7b8399/api/xds_protocol.rst#subscribing-to-resources), it must be that all of those names are unique for the resource type.  Therefore, it does not matter which one of the resource's names is in the `name` field and which are in the `aliases` field.

VHDS is imposing the requirement that the resource must have names corresponding to the RouteConfiguration name and the virtual host name, for all virtual host names associated with the resource.  This requirement is inherent to VHDS and is true regardless of whether we populate the `name` or `aliases` field when the requested resource is unknown.

The only other flexibility here is that xDS servers can provide an additional name for this resource, which may be structured in some other form.  But given that all names of all resources must be unique for that resource type, it really does not matter if that additional name is in the `name` or `aliases` fields.

If the client requests a resource with a given name and that resource does not exist, we know that that resource does not exist, and putting the requested name in the `name` field is completely equivalent to putting it in the `aliases` field.  But since there will not be any aliases in this case, I think it makes sense to just put the requested name in the `name` field.

> This is close to how on-demand request misses are handled right now: the aliases field is used instead of the resource name field, however,

I don't think that VHDS should be special here.  I think that for any incremental xDS method (i.e., any method using `DeltaDiscoveryRequest` and `DeltaDiscoveryResponse`), if the client requests a resource that does not exist, the server should reply by setting the `name` field to the requested name and leaving the resource field unset.  I don't see any reason to use the `aliases` field in this case.

This seems like another detail that should be clarified in the xDS spec doc.  I'd like input from @htuch on this.

Reply:
> The incremental xDS protocol has already established the fact that resources can have more than one name (all but one are called aliases instead of names, but they all serve the same function)

This is not true, at least not for VHDS. Aliases are not names, they potentially are transient, unlike names: vhost domains can change without vhost name changing. I also question the utility of having multiple primary IDs.

To me a resource name is a constant unque ID that can used throughout Envoy. An alias is a temporary resource id, valid from a start of an on-demand request and until the next on-demand request/spontaneous update arrives, at which point any of the aliases can change.

Reply:
The incremental protocol spec does not say that aliases are temporary.  The only thing it really says about them is that "[b]oth the names and aliases of resources should be checked in order to determine whether the entity in question has been subscribed to", which means that names and aliases should be treated the same.  I don't see anything in the spec that says that names and aliases are to be treated differently.

Yes, it's true that virtual hosts can change such that a domain is added or removed.  But I don't see why we need to treat aliases specially in order to support that.  This should work fine without that assumption:

- If a client subscribes to a virtual host resource that does not exist, the server should send an empty message.

- If that domain is then added to an existing virtual host (or a new virtual host is created with that name), the server should send the resource to the client, since it had already subscribed to that name, and the client should detect it by checking for a match on either the `name` or the `aliases` fields.

- If a client is subscribed to a virtual host resource using a domain name that is subsequently removed from the resource (or the resource is removed completely), the server should send the client an empty resource with the `name` field set to the name that the client is subscribed to.

To say this another way, the xDS protocol is about the client subscribing to names, and aliases are about giving a resource multiple names.  The server's responses should be about what names currently exist.

I'd like to hear @htuch's input on this, since this is really not a VHDS-specific question; it's a question about the semantics of the incremental xDS protocol for any resource type.

Reply:
Yeah, my take on this is that I was envisaging name to be the canonical name and aliases to be alternatives, rather than an equivalence set of name strings for a resource, with anyone interchangeable with the other.

Why is this important? I'm actually not sure. I think it could be useful to have a canonical name to simplify comparison of resources, or to make it easier to config dump and have a deterministic result. There are probably other good reasons for this that escape me right now.

I think that for virtual hosts, there might not actually be a sensible canonical name in general, it is really just arbitrary. 

What about having the resource name just be some random opaque string and virtual hosts always be expressed as aliases? Have we considered something like that?

I would be interested in understanding whether we do want just an equivalence set of name in UDPA and if there is a strong case to build that into UDPA-TP.

Reply:
Harvey, I think the core question is the following: Forget about VHDS specifically, and instead think about the incremental xDS protocol semantics for any arbitrary resource type.  In the incremental xDS protocol, when a client is subscribed to a given resource name, what is the right way for the server to tell the client that the specified resource does not exist?

The request just specifies the name the client is interested in, and the protocol requires that it can match on either the `name` or `aliases` fields, which means all such names must be unique for the resource type.  Given that, we know that the requested resource name is not present in any existing resource in either the `name` or `aliases` fields, so sending a reply saying that the `name` does not exist seems completely accurate.  What's the advantage of doing this via the `aliases` field instead?

> Why is this important? I'm actually not sure. I think it could be useful to have a canonical name to simplify comparison of resources, or to make it easier to config dump and have a deterministic result. There are probably other good reasons for this that escape me right now.

If the server needs a canonical name for comparisons or simplifying config dumps, it can prefer the one in the `name` field over the ones in the `aliases` field.  But that is an implementation detail for the server, not something that should affect the protocol between the client and server.

If we can't think of good reasons to treat the aliases as second-class citizens, then I don't think we should bake that into the protocol.  I don't see anything in the current protocol that says that they should be treated differently in any way.

> What about having the resource name just be some random opaque string and virtual hosts always be expressed as aliases? Have we considered something like that?

VHDS requires that the VirtualHost resource must have at least one name (whether it be in the `name` or `aliases` field) for each domain, so that clients can query for them.  However, it does not place any restrictions on any other names that may be used for the resource, so the server is free to assign any additional name(s) it may want.

It's not clear to me why the server would want to assign additional name(s), but I have no objection to it doing so.  I don't even have any objection to one of those additional names being the one in the `name` field instead of being one of the ones in the `aliases` field, since I think we should be treating all names as equivalent.

But, even if the server does set the `name` field to its own internal name for all VirtualHost resources that *do* exist, I still think that the incremental xDS protocol should use the `name` field to indicate requested resources that do *not* exist, as per the reasoning described above.

> I would be interested in understanding whether we do want just an equivalence set of name in UDPA and if there is a strong case to build that into UDPA-TP.

For UDPA, unless we can come up with specific use-cases where we would need to treat names and aliases differently, I would actually argue to replace the separate `name` and `aliases` fields with a single repeated `name` field.  Servers that need a canonical name for things like comparisons or config dumps can just use the first name in the list.

Reply:
I've created #9317 to attempt to clarify the incremental xDS protocol semantics for resources that do not exist.

Reply:
Regardless, this is how it's implemented. Based on this conversation, it seems that @htuch @dmitri-d and myself are in agreement that this is good for now. Can we address this after this separately from this PR?

Reply:
Sure, let's take #9317 and see where this leads us. I feel that without another on-demand protocol, we don't have much in the way of examples to illustrate the utility of a canonical name, we can see how this plays out in v2 xDS to feed into UDPA.

Reply:
I don't think that we can both accept #9317 *and* say that VHDS detects unknown virtual hosts based on the `aliases` field instead of the `name` field.  This is basically a contradiction.  It would mean that in practice, every management server would need to do something in violation of the spec.

I think we have two choices here:

1. Fix VHDS to check the `name` field instead of `aliases` field, and update this PR accordingly.
2. Change #9317 to say that clients need to be prepared to match unknown resources on both the `name` and `aliases` fields.

I would prefer option (1), since option (2) would impose additional requirements on all client implementations without actually adding any value.

Reply:
The code is already implemented as described in this doc. 9317 is proposing to change that. Whether that's right or wrong, I'm not sure. However, what's documented here is current functionality and agreed upon by @htuch @dmitri-d and myself. I'm proposing pushing protocol and docs changes to the other PR and we can update if we add these changes if we need to.

Reply:
It seems #9317 is a very recent PR and an attempt to change the requirements for work which has already been done and agreed to, this should ideally be handled through other PRs after this work is committed.  IMHO this has been an enormous task and we need to draw a line under it so we can move forward with it being the basis for future changes.